### ISSUE TITLE
CD-ROM library

### ISSUE THREAD

**Issue by mihaip** [mihaip/infinite-mac#166]
CD-ROM games (Myst, etc.) do work when loaded as disk images, but it's somewhat awkward to have to download them, drag them in, and do a restart (until #49 is fixed anyway).

Ideally there should be a browser UI (outside the simulator, styled as a CD wallet?) that lets you browse and load things directly from the site (or via a simple proxy if CORS is not available).

- [x] Nicer headers
- [x] Increase Infinite HD size
- [x] Left-align CD-ROMs tab at narrow widths so that it doesn't overlap with the footer
- [ ] macintoshgarden.org URLs fail in production (Cloudflare may be blocked)
- [x] Custom URL option in library
- [x] Search-as-you-type filter
- [x] Make it work in Mini vMac (may need to turn off [non-disk image protection](https://www.gryphel.com/c/minivmac/options.html#option_ndp))
- [x] Get actual disk size by reading the partition header (so that [truncated images](https://img.classicmacdemos.com/multi-pong.img) work)
- [ ] #204
- [x] #205
- [ ] Things to add to the library
    - [x] Things from [this list of early games](http://cdrom.ca/games/2023/06/29/first-cdrom-games.html)
    - [x] Riven (current copy is broken)
    - [x] The Journeyman Project
    - [ ] Lunicus
    - [ ] Marathon Trilogy (the CD with all of the bonus content)
    - [ ] 11th Hour
    - [ ] More Apple Developer CDs?
    - [x] MacAddict "The Disc"
    - [ ] [AfterDark compilations](https://archive.org/details/after-dark-4-deluxe)
    - [ ] Grim Fandango
    - [ ] Full Throttle
    - [x] Things [archived by mac84tv](https://twitter.com/Mac84tv/status/1656115643161223169)
    - [x] Move "Just Grandma and Me" to from Library to CD-ROMs
    - [x] Myth
    - [ ] Where In the World Is Carmen San Diego?
    - [ ] The Last Express
    - [ ] Power Pete
    - [x] https://bitbang.social/@mac84tv/110746928981347355
    - [ ] Dazzeloids
    - [ ] Additional Voyager titles ([thread](https://hachyderm.io/@peterme@sfba.social/110825738324996612), may require Red Book/Audio CD support for some)
        - [ ] Igor Stravinsky: The Rite of Spring
        - [ ] Ludwig van Beethoven: Symphony No. 9
        - [ ] Who Built America?
        - [ ] The Day After Trinity
    - [ ] [MachTen](https://archive.org/details/power-machten414)
- [x] Move stickies so that they don't cover up the mounted disks
- [x] Features that this enables
   - [x] #48 
   - [x] #87

**Comment by jcgraybill** [comment]
:wave:, I'm working to set up some disk images for https://classicmacdemos.com/ that could work with this feature. I notice that even though the "CD" in the name of the feature/issue implies this is for .iso files, [this link to an .img file](https://infinitemac.org/1996/System%207.5.3?cdroms=true&cdrom_url=https%3A%2F%2Farchive.org%2Fdownload%2Fmac-demo-games%2FMacDemoGames.img) on Internet Archive works just fine. 

It'd work best in my workflow to create .dmg files for these, but before I set off AppleScripting that together, I want to confirm whether that disk image format is one you're actually intending to support. Let me know and I'll push a couple of test images to the site's CDN to try out. Thanks!

**Comment by mihaip** [comment]
.dmg is harder (since it's compressed, and not natively supported by the Linux build of Basilisk II/SheepShaver). 

For now I'm only supporting raw .iso images (I'm not 100% clear on what the difference between .iso and .img is -- it seems to be somewhat arbitrary as to which is used).

**Comment by that-ben** [comment]
It's also worth noting that .img can depict a dozen different disk image formats... and at least one of them cannot not even exist outside of a HFS volume since it requires a resource fork.  In the classic Mac world, a .img file is usually a Disk Copy image, though some people wrongly name dd images as .img while they should (at least IMO) name them .dsk to avoid all possible confusion.  In the Windows world, .img is typically used for floppy disk images though it's also used for other types of media.

TL;DR it's quite a mess and one should avoid naming any disk image .img

Also, it's funny that most Mac .iso files are actually not even ISO disk images.

**Comment by jcgraybill** [comment]
Thanks @mihaip and @that-ben. Wow, I am finding it's quite a challenge to create .iso disk images in the old Mac CD format on a modern computer. I've created a few sample images in different formats and tested mounting them with various Infinite Mac VMs. Here's what I'm finding. Feel free to download and try these images yourself.

[Uncompressed HFS+ .img](https://download.classicmacdemos.com/infinitemac/diskcopy-hfs-plus.img): The easiest to create, straight out of Disk Copy. Works on Mac OS 8.1+ VMs due to HFS+ support. I also converted this into two CD-ROM formats, and these worked equally well. If all three work, I'd prefer to use .img images and avoid a conversion step.

- [Uncompressed HFS+ .cdr](https://download.classicmacdemos.com/infinitemac/diskcopy-hfs-plus.cdr): Converted with `hdiutil convert -format UDTO -o diskcopy-hfs-plus.cdr diskcopy-hfs-plus.img`.
- [Uncompressed HFS+ .iso](https://download.classicmacdemos.com/infinitemac/diskcopy-hfs-plus.cdr): Converted with `hdiutil makehybrid -iso -joliet -hfs -o diskcopy-hfs-plus.iso diskcopy-hfs-plus.img`.

[Uncompressed HFS .img](https://download.classicmacdemos.com/infinitemac/diskcopy-hfs.img): Created with DiskCopy on Mac OS 7.6. This mounts on all System 7.0+ VMs, and mounts (successfully) with a warning that it "needs minor repairs" on System 2.1 through System 6.0.5. While this is more compatible, I'd probably reserve it for games that don't work on Mac OS 8, since my workflow right now is all based on Mac OS 9.
- I also created an [Uncompressed HFS .cdr](https://download.classicmacdemos.com/infinitemac/diskcopy-hfs.cdr) with `hdiutil convert -format UDTO -o diskcopy-hfs.cdr diskcopy-hfs.img`. Just like with the HFS+ images, this didn't appear to be any less or more compatible in the testing I performed. Diskutil can't make a hybrid/ISO image from HFS, so I didn't create one of those.

Based on this testing, I'd propose that I create uncompressed HFS / HFS+ .img files with Disk Copy. But let me know if there's value in converting these to .cdr or .iso for any benefits these tests wouldn't have shown.

**Comment by that-ben** [comment]
By "Disk Copy" I take it you meant "Disk Utility" correct? Disk Copy 6 images are not binary safe formats and will not work when moved to anything except a Mac partition, including moving it to an web server or any Windows drive such as FAT, exFAT, or NTFS.  Only Disk Copy 4 can produce a binary safe .img format.

**Comment by jcgraybill** [comment]
Okay that is... amazing! 

I realize I sort of started in the middle of things, no surprise it's a bit confusing to read. The short version is, I'm using a Mac OS 9.2.1 QEMU virtual machine to grab each demo from its coverdisc, take screenshots, extract the application icon and README, and create a binhexed Stuffit archive, which is what end users download from [the website](https://www.classicmacdemos.com/). (For games that won't run in an emulator, I'm using an actual G4 Macintosh for this step...) I transfer everything briefly from the QEMU VM to a modern macOS machine to upload it to CloudFront/S3. I like Stuffit as a format for end-users, because Stuffit Expander is so widely available, and the archive extracts to a folder on the user's drive. 

I want to also create disk images that can be auto-mounted in an Infinite Mac VM. The idea is that end users will never interact with these: they'll be passed as a parameter to an Infinite Mac launch URL, and will be mounted as a drive in the VM. So my priority for these are, maximum future compatibility with the Infinite Mac project, and the ability to hook into the existing workflow - which could either be on the Mac OS 9 or the macOS stage. For that, Disk Copy 6 is pretty compelling if the .img images it creates do indeed work with Infinite Mac.

**Comment by that-ben** [comment]
If it's uncompressed, unencrypted it could theoretically work, but if it's compressed, it won't work.  Here's an example I just made of a compressed DiskCopy 6.4 image.  That doesn't work in Infinite Mac:
[MR Browser DC 6.4 ROC.dsk.zip](https://github.com/mihaip/infinite-mac/files/11951636/MR.Browser.DC.6.4.ROC.dsk.zip)

**Comment by jcgraybill** [comment]
Nice - that's what I'm seeing too. Your example image didn't work, but the uncompressed DiskCopy 6.3.3 images ☝️ mount just fine in Infinite Mac.

**Comment by jcgraybill** [comment]
Okay, I've placed a test image file here: https://img.classicmacdemos.com/abuse.img

I *believe* I've configured CORS on this CloudFront distribution to allowlist infinitemac.org and www.infinitemac.org. If I got everything right, then this URL should open an InfiniteMac VM with the Abuse image mounted. @mihaip you would just need to allowlist the img.classicmacdemos.com domain.

https://infinitemac.org/1998/Mac%20OS%208.1?machine=Power+Macintosh+9500&cdroms=true&cdrom_url=https://img.classicmacdemos.com/abuse.img

**Comment by that-ben** [comment]
I think you need to URL encode your cdrom_url parameter like this: https%3A%2F%2Fimg.classicmacdemos.com%2Fabuse.img

**Comment by jcgraybill** [comment]
> I think you need to URL encode your cdrom_url parameter like this: https%3A%2F%2Fimg.classicmacdemos.com%2Fabuse.img

I originally thought so too, but I discovered that url-encoding the cdrom_url parameter doesn't seem to be required. E.g. https://infinitemac.org/1998/Mac%20OS%208.1?machine=Power+Macintosh+9500&cdroms=true&cdrom_url=https://archive.org/download/macaddict-cd-055/MacAddict_055_2001_03.iso works just fine.

Happy to do it if that's preferred, though.

**Comment by mihaip** [comment]
@jcgraybill the allowlist is now updated, those URLs should work now.

**Comment by jcgraybill** [comment]
@mihaip It works great! This is so nifty, I'll add it to several of the demos right now.

**Comment by jcgraybill** [comment]
Here's how it shows up on the website. https://www.classicmacdemos.com/multi-pong

Thanks @mihaip and @that-ben ! People are going to love this.

**Comment by that-ben** [comment]
Very nice! It's pretty hard a game though!!!

**Comment by jcgraybill** [comment]
> Very nice! It's pretty hard a game though!!!

Lol that's my one easter egg on the site, the game I made last summer. Even I can almost never beat it on normal difficulty. :)

**Comment by jcgraybill** [comment]
Interestingly, I'm noticing today that .img images are causing an emulator error ("RangeError: source array is too long") when they load. E.g. https://infinitemac.org/1998/Mac%20OS%208.1?machine=Power+Macintosh+9500&cdroms=true&cdrom_url=https://img.classicmacdemos.com/multi-pong.img

**Comment by mihaip** [comment]
The file size of https://img.classicmacdemos.com/multi-pong.img is 28K, but it's actually a 1MB disk (I assume there's an optimization in the format to omit empty space at the end). For now I assume that the disk size is the same as the file size, which is leading to an out-of-bounds write.

I'll see if it's possible to get the actual disk size by reading a bit of the disk image when getting the metadata.

**Comment by that-ben** [comment]
So effectively, this is a truncated (very crude compression technique) disk image.  Indeed, it would be interesting if you can come up with reading/decoding the bytes that make up for the real volume size.

**Comment by jcgraybill** [comment]
Interesting, this must be a DiskCopy trick - these come straight out of DiskCopy 6.3.3, and I can see that it's indeed rounding up the reported volume size to 1MB. 

<img width="723" alt="Screenshot 2023-07-16 at 7 39 43 AM" src="https://github.com/mihaip/infinite-mac/assets/65628360/10915a95-56b0-4555-beb7-fdd6dc8da70a">

FWIW, I can confirm these images were mounting correctly into Infinite Mac emulators last week, so I wonder if these are getting caught by a recently-introduced change.

**Comment by that-ben** [comment]
That is also why I don't recommend using Disk Copy 6 to make any kind of disk image.  Disk Copy 4.2 or ShrinkWrap DiskDup+ is the way to go.  It makes universally compatible disk images.  Disk Copy 6 does not.

**Comment by jcgraybill** [comment]
Yes, definitely open to converting to DiskDup+ images, if that is easier to work with.

**Comment by mihaip** [comment]
> FWIW, I can confirm these images were mounting correctly into Infinite Mac emulators last week, so I wonder if these are getting caught by a recently-introduced change.

Ah, good observation. I used to mount everything as read-only, but given that they can be boot disks, doing it as read/write seems more correct. This is leading to the Finder doing a small write when the image is mounted (possibly a desktop database rebuild).

I've fixed this, and https://infinitemac.org/1998/Mac%20OS%208.1?machine=Power+Macintosh+9500&cdrom_url=https://img.classicmacdemos.com/multi-pong.img should load now.

**Comment by that-ben** [comment]
It does! :+1: 

![adsdas](https://github.com/mihaip/infinite-mac/assets/13210399/6359e0c6-6001-4e0a-96bf-6bc0297e4f2d)

**Comment by jcgraybill** [comment]
Hot! We're back in business. I re-enabled Infinite Mac links on the website, and now folks can again play Quake in their web browser with a single click. :) 

https://www.classicmacdemos.com/quake
